File: src/layers/auth-layer/python/auth_attempt.py
import auth_constants
import common
import constants
from boto3.dynamodb.conditions import Attr
from common_marshmallow import BaseSchema
from marshmallow import ValidationError, fields, pre_load, validates_schema


class AuthAttemptDbSchema(BaseSchema):
    pk = fields.String(required=True)
    sk = fields.Constant("Attempt")
    action = fields.String(required=True)
    ip_addr = fields.String()
    jti = fields.String()
    attempt = fields.Integer(required=True)
    expired_at = fields.Integer()

    @pre_load
    def gen(self, data, **kwargs):
        action = common.convert_snake_case_to_pascal_case(src=data["action"])

        if data.get("ip_addr"):
            data["pk"] = f"{action}#{data['ip_addr']}"

        elif data.get("jti"):
            data["pk"] = f"{action}#{data['jti']}"

        return data

    @validates_schema
    def validate(self, data, **kwargs):
        if not data.get("ip_addr") and not data.get("jti"):
            raise ValidationError("Either ip_addr or jti must exist")


class AuthAttemptDb:
    def __init__(self, dynamodb, table):
        self.dynamodb = dynamodb
        self.table = table

    def get(self, action: str, ip_addr="", jti="") -> dict:
        action = common.convert_snake_case_to_pascal_case(src=action)

        db_resp = self.table.get_item(
            Key={"pk": f"{action}#{ip_addr or jti}", "sk": "Attempt"},
            ProjectionExpression="attempt,expired_at",
        )

        return db_resp.get("Item", {})

    def is_blocked(self, action: str, ip_addr="", jti="") -> bool:
        attempt = self.get(action=action, ip_addr=ip_addr, jti=jti)

        return bool(
            attempt
            and (attempt.get("expired_at") or constants.MAX_TIMESTAMP_IN_SECONDS)
            > common.get_current_timestamp()
            and attempt["attempt"] >= auth_constants.MAX_ACTION_ATTEMPT_DICT[action]
        )

    def incr(self, action: str, ip_addr="", jti="", attempt=1):
        expired_at = common.extend_current_timestamp(hours=1)

        try:
            self.table.update_item(
                Key={
                    "pk": f"{common.convert_snake_case_to_pascal_case(src=action)}#{ip_addr or jti}",
                    "sk": "Attempt",
                },
                UpdateExpression="SET attempt = attempt + :incr, expired_at = :ea",
                ConditionExpression=Attr("pk").exists(),
                ExpressionAttributeValues={":incr": attempt, ":ea": expired_at},
            )

        except self.dynamodb.meta.client.exceptions.ConditionalCheckFailedException:
            attempt_item = {
                "action": action,
                "attempt": attempt,
                "expired_at": expired_at,
            }

            if ip_addr:
                attempt_item["ip_addr"] = ip_addr

            if jti:
                attempt_item["jti"] = jti

            self.table.put_item(
                Item=AuthAttemptDbSchema().load_and_dump(attempt_item),
                ConditionExpression=Attr("pk").not_exists(),
            )

    def block(self, action: str, ip_addr="", jti="", is_permanent=False):
        expired_at = common.extend_current_timestamp(
            minutes=(
                auth_constants.JWT_TOKEN_VALIDITY_IN_MINUTES_DICT[action] if jti else 60
            )
        )

        try:
            self.table.update_item(
                Key={
                    "pk": f"{common.convert_snake_case_to_pascal_case(src=action)}#{ip_addr or jti}",
                    "sk": "Attempt",
                },
                UpdateExpression="SET attempt = :a, expired_at = :ea",
                ConditionExpression=Attr("pk").exists(),
                ExpressionAttributeValues={
                    ":a": auth_constants.MAX_ACTION_ATTEMPT_DICT[action],
                    ":ea": None if is_permanent else expired_at,
                },
            )

        except self.dynamodb.meta.client.exceptions.ConditionalCheckFailedException:
            attempt_item = {
                "action": action,
                "attempt": auth_constants.MAX_ACTION_ATTEMPT_DICT[action],
                "expired_at": expired_at,
            }

            if ip_addr:
                attempt_item["ip_addr"] = ip_addr

            if jti:
                attempt_item["jti"] = jti

            self.table.put_item(
                Item=AuthAttemptDbSchema().load_and_dump(attempt_item),
                ConditionExpression=Attr("pk").not_exists(),
            )

File: src/auth-lib/src/auth_attempt.rs
use crate::{auth_constants, auth_enums::Action};
use anyhow::{Context, Error, Result};
use aws_sdk_dynamodb::{
    error::SdkError, operation::update_item::UpdateItemError::ConditionalCheckFailedException,
    types::AttributeValue,
};
use common::{self, Case};
use optarg2chain::optarg_impl;
use serde::{Deserialize, Serialize};
use std::panic::Location;

#[derive(
    Clone, Copy, Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize,
)]
enum Sk {
    #[default]
    Attempt,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct AuthAttempt {
    pk: String,

    #[serde(default)]
    sk: Sk,

    action: Action,

    #[serde(default, skip_serializing_if = "String::is_empty")]
    ip_addr: String,

    #[serde(default, skip_serializing_if = "String::is_empty")]
    jti: String,

    attempt: u32,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    expired_at: Option<u64>,
}

#[optarg_impl]
impl AuthAttempt {
    #[optarg_method(AuthAttemptNewBuilder, call)]
    pub fn new(
        action: Action,
        #[optarg_default] ip_addr: String,
        #[optarg_default] jti: String,
        attempt: u32,
        expired_at: u64,
    ) -> Self {
        let pk_action = action.to_string().convert_snake_case_to_pascal_case();

        Self {
            pk: if ip_addr.is_empty() {
                format!("{pk_action}#{jti}")
            } else {
                format!("{pk_action}#{ip_addr}")
            },
            sk: Sk::Attempt,
            action,
            ip_addr,
            jti,
            attempt,
            expired_at: Some(expired_at),
        }
    }
}

#[derive(Debug)]
pub struct AuthAttemptDb<'a> {
    pub dynamodb: &'a aws_sdk_dynamodb::Client,
}

#[optarg_impl]
impl<'a> AuthAttemptDb<'a> {
    #[optarg_method(AuthAttemptDbGetItemBuilder, send)]
    pub async fn get_item<'b>(
        &'a self,
        action: Action,
        #[optarg_default] ip_addr: &'b str,
        #[optarg_default] jti: &'b str,
    ) -> Result<Option<AuthAttempt>> {
        let db_resp = self
            .dynamodb
            .get_item()
            .table_name(&*auth_constants::AUTH_ATTEMPT_TABLE_NAME)
            .key(
                "pk",
                AttributeValue::S(format!(
                    "{action}#{id}",
                    action = action.to_string().convert_snake_case_to_pascal_case(),
                    id = if ip_addr.is_empty() { jti } else { ip_addr }
                )),
            )
            .key("sk", AttributeValue::S("Attempt".to_string()))
            .projection_expression("pk,#action,attempt,expired_at")
            .expression_attribute_names("#action", "action")
            .send()
            .await
            .context(Location::caller())?;

        if let Some(item) = db_resp.item {
            Ok(serde_dynamo::from_item(item).context(Location::caller())?)
        } else {
            Ok(None)
        }
    }

    #[optarg_method(AuthAttemptDbIsBlockedBuilder, send)]
    pub async fn is_blocked<'b>(
        &'a self,
        action: Action,
        #[optarg_default] ip_addr: &'b str,
        #[optarg_default] jti: &'b str,
    ) -> Result<bool> {
        if let Some(attempt) = self
            .get_item(action)
            .ip_addr(ip_addr)
            .jti(jti)
            .send()
            .await
            .context(Location::caller())?
        {
            let is_blocked = attempt.expired_at.unwrap_or(u64::MAX)
                > common::get_current_timestamp()
                    .call()
                    .context(Location::caller())?
                && attempt.attempt >= auth_constants::MAX_ACTION_ATTEMPT_MAP[&action];

            Ok(is_blocked)
        } else {
            Ok(false)
        }
    }

    #[optarg_method(AuthAttemptDbIncrBuilder, send)]
    pub async fn incr<'b>(
        &'a self,
        action: Action,
        #[optarg_default] ip_addr: &'b str,
        #[optarg_default] jti: &'b str,
        #[optarg(1)] attempt: u32,
        #[optarg_default] is_permanent: bool,
    ) -> Result<()> {
        let extend_by_in_minutes = if jti.is_empty() {
            60u64
        } else {
            auth_constants::JWT_TOKEN_VALIDITY_IN_MINUTES_MAP[&action]
        };

        let expired_at = common::extend_current_timestamp()
            .minutes(extend_by_in_minutes)
            .call()
            .context(Location::caller())?;

        let db_resp = self
            .dynamodb
            .update_item()
            .table_name(&*auth_constants::AUTH_ATTEMPT_TABLE_NAME)
            .key(
                "pk",
                AttributeValue::S(format!(
                    "{action}#{id}",
                    action = action.to_string().convert_snake_case_to_pascal_case(),
                    id = if ip_addr.is_empty() { jti } else { ip_addr }
                )),
            )
            .key("sk", AttributeValue::S("Attempt".to_string()))
            .update_expression("SET attempt = attempt + :incr, expired_at = :ea")
            .condition_expression("attribute_exists(pk)")
            .expression_attribute_values(":incr", AttributeValue::N(attempt.to_string()))
            .expression_attribute_values(
                ":ea",
                if is_permanent {
                    AttributeValue::Null(true)
                } else {
                    AttributeValue::N(expired_at.to_string())
                },
            )
            .send()
            .await
            .context(Location::caller());

        if let Err(err) = db_resp {
            let err = err
                .downcast::<SdkError<_>>()
                .context(Location::caller())?
                .into_service_error();

            if let ConditionalCheckFailedException(_) = err {
                let attempt = AuthAttempt::new(action, attempt, expired_at)
                    .ip_addr(ip_addr)
                    .jti(jti)
                    .call();

                self.dynamodb
                    .put_item()
                    .table_name(&*auth_constants::AUTH_ATTEMPT_TABLE_NAME)
                    .set_item(Some(
                        serde_dynamo::to_item(attempt).context(Location::caller())?,
                    ))
                    .condition_expression("attribute_not_exists(pk)")
                    .send()
                    .await
                    .context(Location::caller())?;
            } else {
                return Err(Error::from(err).context(Location::caller()));
            }
        }

        Ok(())
    }
}
